velox/exec/Merge.cpp

```diff
@@ -165,60 +200,186 @@ RowVectorPtr Merge::getOutput() {
+    bool& lastRun) {
+  VELOX_CHECK_NOT_NULL(merger_);
+  if (!output_) {
+    output_ = BaseVector::create<RowVector>(type_, maxOutputRows, pool_);
+    for (auto& child : output_->children()) {
+      child->resize(maxOutputRows);
+    }
+  }
+
+  uint64_t outputSize = 0;
+  for (;;) {
+    const auto stream = merger_->next();
+    if (stream == nullptr) {
+      if (outputSize > 0) {
+        output_->resize(outputSize);
+      }
+      lastRun = true;
       return std::move(output_);
     }
 
-    if (stream->setOutputRow(outputSize_)) {
-      // The stream is at end of input batch. Need to copy out the rows before
-      // fetching next batch in 'pop'.
+    if (stream->setOutputRow(outputSize)) {
+      // The stream is at end of input batch. Need to copy out the rows
+      // before fetching next batch in 'pop'.
       stream->copyToOutput(output_);
     }
 
-    ++outputSize_;
+    ++outputSize;
 
     // Advance the stream.
-    stream->pop(sourceBlockingFutures_);
+    stream->pop(sourceBlockingFutures);
 
-    if (outputSize_ == outputBatchSize_) {
+    if (outputSize == maxOutputRows) {
       // Copy out data from all sources.
-      for (auto& s : streams_) {
+      for (const auto& s : streams_) {
         s->copyToOutput(output_);
       }
-
-      outputSize_ = 0;
       return std::move(output_);
     }
 
-    if (!sourceBlockingFutures_.empty()) {
+    if (!sourceBlockingFutures.empty()) {
       return nullptr;
     }
   }
 }
 
-void Merge::close() {
-  for (auto& source : sources_) {
-    source->close();
+std::unique_ptr<TreeOfLosers<SpillMergeStream>> SpillMerger::createSpillMerger(
+    std::vector<std::vector<std::unique_ptr<SpillReadFile>>>
+        spillReadFilesGroup) {
+  std::vector<std::unique_ptr<SpillMergeStream>> streams;
+  streams.reserve(spillReadFilesGroup.size());
+  for (auto i = 0; i < spillReadFilesGroup.size(); ++i) {
+    streams.push_back(ConcatFilesSpillMergeStream::create(
+        i, std::move(spillReadFilesGroup[i])));
+  }
+  return std::make_unique<TreeOfLosers<SpillMergeStream>>(std::move(streams));
+}
+
+RowVectorPtr SpillMerger::getOutput(vector_size_t maxOutputRows) {
+  // Finished.
+  if (numOutputRows_ == numSpilledRows_) {
+    return nullptr;
   }
+
+  VELOX_CHECK_GT(maxOutputRows, 0);
+  VELOX_CHECK_GT(numSpilledRows_, numOutputRows_);
+  const vector_size_t batchSize =
+      std::min<uint64_t>(numSpilledRows_ - numOutputRows_, maxOutputRows);
+  if (output_ != nullptr) {
+    VectorPtr output = std::move(output_);
+    BaseVector::prepareForReuse(output, batchSize);
+    output_ = std::static_pointer_cast<RowVector>(output);
+  } else {
+    output_ = std::static_pointer_cast<RowVector>(
+        BaseVector::create(type_, batchSize, pool_));
+  }
+
+  for (const auto& child : output_->children()) {
+    child->resize(batchSize);
+  }
+
+  spillSources_.resize(batchSize);
+  spillSourceRows_.resize(batchSize);
+
+  VELOX_CHECK_GT(output_->size(), 0);
```

Comment:
I assume the code logic here is copied from existing spill code? Thanks!

Reply:
Yes